{"metadata": {"language_info": {"name": "bash", "codemirror_mode": "shell", "mimetype": "text/x-sh", "file_extension": ".sh"}, "kernelspec": {"name": "bash", "display_name": "Bash", "language": "bash"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# Run regenie: Resistant hypertension replication\n\nThis script preprocesses genotype data and runs GWAS with SNP data. We replicate analysis performed earlier with FinnGen data. Both analysis steps will be carried out with SNP data as was done in FinnGen.\n\nPhenotypes have been created by script **create_ukbb_pheno_resht.Rmd**\n\n**Variable definitions**\n\n* Genotypes: \n  - 22418, chip data (GRCh37) for regenie step 1\n  - 21008, imputed data, Genomics England (GRCh38) for regenie step2\n* Phenotypes: \n  - resht_fg: Number of hypertensive medication classes $\\geq$ 4 and hypertension\n  - resht_ukb: Number of medication classes $\\geq$ 3, Systolic BP  $\\geq$ 140 or diastolic BP  $\\geq$ 90 and hypertension\n  - resht_both: resht_fg or resht_ukbb\n  - Control for all variables: One hypertensive medication class and hypertension\n * Covariates:\n   - 21003, Age when attended assessment centre (Baseline characteristics)\n   - 22001, Sex (Genotypes)\n   - 22009, Genetic principal components (Genotypes)\n   \n**Run variables**", "metadata": {"tags": []}}, {"cell_type": "code", "source": "name=\"resht\"\npheno_file_dir=\"/data\"\ngeno_file_dir=\"/data/genotype\"\nregenie_out_dir=\"/data/regenie\"\n\ngeno_var=\"ukb22418\"\ngeno_in=\"/mnt/project/data/lifted_geno_in/${geno_var}_c*.???\"  #[1-9]*\"\nbed_base=\"${geno_var}_c1_22_merged\" #_c10_b0_v2_GRCh38_full_analysis_set_plus_decoy_hla.fam  \n\ngeno2_var=\"ukb21008\"\ngeno2_indir=\"/Bulk/Imputation/Imputation\\ from\\ genotype\\ \\(GEL\\)\"\n\npheno_in=\"resht_pheno.tsv\"\npheno_cols=\"resht_fg,resht_ukb,resht_both\"\ncovar_cols=\"p21003,p22001,p22009_a1,p22009_a2,p22009_a3,p22009_a4,p22009_a5,p22009_a6,p22009_a7,p22009_a8,p22009_a9,p22009_a10\"\n\ninstance=\"mem1_ssd1_v2_x16\"", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "markdown", "source": "## Lift over\n   \nThe chip data is in the GRCh37 reference, while imputed data is in the GRCh37 reference. We will lift over the chip data by liftover_plink_beds pipeline: https://github.com/dnanexus-rnd/liftover_plink_beds\\n\". The protocol is described in the file `run_liftover.ipynb`. ", "metadata": {"tags": []}}, {"cell_type": "markdown", "source": "## Preprocess genotype data\n   \n### Chip data\n\n#### Merge chromosome data  ", "metadata": {"tags": []}}, {"cell_type": "code", "source": "run_merge=\"cp ${geno_in} . ;\\\n           ls *.bed | sed -e 's/.bed//g' > files_to_merge.txt;\\\n           plink --merge-list files_to_merge.txt \\\n                 --make-bed \\\n                 --chr 1-22 \\\n                 --allow-extra-chr \\\n                 --out ${bed_base};\\\n           rm files_to_merge.txt;\"\n             \ndx run swiss-army-knife \\\n           -icmd=\"${run_merge}\" \\\n           --tag=\"Step0\" \\\n           --instance-type ${instance} \\\n           --destination=\"${geno_file_dir}/\" \\\n           --brief --yes --wait\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [{"name": "stdout", "text": "job-Gb74yZQJFFQbp5xgzJxgzg3q\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "#### Filter genotypes", "metadata": {"tags": []}}, {"cell_type": "code", "source": "run_plink_qc=\"plink2 --bfile ${bed_base} \\\n                     --keep ${pheno_in}  \\\n                     --autosome \\\n                     --maf 0.01 --mac 20 --geno 0.1 \\\n                     --hwe 1e-15 --mind 0.1 \\\n                     --write-snplist \\\n                     --write-samples \\\n                     --no-id-header \\\n                     --out geno_qc_pass\"\n\ndx run swiss-army-knife \\\n   -iin=\"${geno_file_dir}/${bed_base}.bed\" \\\n   -iin=\"${geno_file_dir}/${bed_base}.bim\" \\\n   -iin=\"${geno_file_dir}/${bed_base}.fam\" \\\n   -iin=\"${pheno_file_dir}/${pheno_in}\" \\\n   -icmd=\"${run_plink_qc}\" \\\n   --tag=\"Step0\" \\\n   --instance-type ${instance} \\\n   --destination=\"${geno_file_dir}/\" \\\n   --brief --yes --wait\n   ", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": [{"name": "stdout", "text": "job-Gb952BjJFFQbv07q75Kxv3G2\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Run Regenie\n\nWe will run with similar settings as in FinnGen.\n\n### Regenie step 1: Fit model", "metadata": {}}, {"cell_type": "code", "source": "run_regenie_step1=\"regenie \\\n  --step 1 \\\n  --bt \\\n  --bed ${bed_base} \\\n  --extract geno_qc_pass.snplist \\\n  --phenoFile ${pheno_in} \\\n  --phenoColList '${pheno_cols}' \\\n  --covarFile ${pheno_in} \\\n  --covarColList '${covar_cols}' \\\n  --bsize 1000 \\\n  --lowmem \\\n  --gz \\\n  --threads 16 \\\n  --out ${name}_step1\"\n\ndx run swiss-army-knife \\\n   -iin=\"${geno_file_dir}/${bed_base}.bed\" \\\n   -iin=\"${geno_file_dir}/${bed_base}.bim\" \\\n   -iin=\"${geno_file_dir}/${bed_base}.fam\" \\\n   -iin=\"${geno_file_dir}/geno_qc_pass.snplist\" \\\n   -iin=\"${pheno_file_dir}/${pheno_in}\" \\\n   -icmd=\"${run_regenie_step1}\" \\\n   --tag=\"Step1\" --instance-type ${instance} \\\n   --destination=\"${regenie_out_dir}/\" \\\n   --brief --yes --wait\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [{"name": "stdout", "text": "job-Gb95PzQJFFQvQbjpBy194G96\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "### Regenie step 2: Association testing\n\nWe perform association testing for SNP data.", "metadata": {}}, {"cell_type": "code", "source": "for chr in {1..22}; do\n\n    run_regenie_step2=\"regenie \\\n      --step 2 \\\n      --bgen ${geno2_var}_c${chr}_b0_v1.bgen \\\n      --sample ${geno2_var}_c${chr}_b0_v1.sample \\\n      --ref-first \\\n      --phenoFile ${pheno_in} \\\n      --phenoColList '${pheno_cols}' \\\n      --covarFile ${pheno_in} \\\n      --covarColList '${covar_cols}' \\\n      --pred ${name}_step1_pred.list \\\n      --bt \\\n      --firth  \\\n      --firth-se\\\n      --approx \\\n      --bsize 400 \\\n      --pThresh 0.05 \\\n      --threads 16 \\\n      --gz \\\n      --out assoc_c${chr}\"\n\n    dx run swiss-army-knife \\\n       -iin=\"${geno2_indir}/${geno2_var}_c${chr}_b0_v1.bgen\" \\\n       -iin=\"${geno2_indir}/${geno2_var}_c${chr}_b0_v1.sample\" \\\n       -iin=\"${geno_file_dir}/geno_qc_pass.snplist\" \\\n       -iin=\"${pheno_file_dir}/${pheno_in}\" \\\n       -iin=\"${regenie_out_dir}/${name}_step1_pred.list\" \\\n       -iin=\"${regenie_out_dir}/${name}_step1_1.loco.gz\" \\\n       -iin=\"${regenie_out_dir}/${name}_step1_2.loco.gz\" \\\n       -iin=\"${regenie_out_dir}/${name}_step1_3.loco.gz\" \\\n       -icmd=\"${run_regenie_step2}\" \\\n       --tag=\"Step2\" --instance-type ${instance} \\\n       --destination=\"${regenie_out_dir}/\" \\\n       --brief --yes\n\ndone\n\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "#Wait until all chromosomes have been run \n# for chr in {1..22}; do\n#   until [ -f /mnt/project${regenie_out_dir}/assoc_c${chr}.log ]\n#   do\n#     sleep 120s  #Check after every 2 min\n#  done  \n#done\n#sleep 120s #Wait 2 min... log f", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Post processing", "metadata": {}}, {"cell_type": "markdown", "source": "### Merge", "metadata": {}}, {"cell_type": "code", "source": "#Perhaps if I make script out of this, I could make this work with SAK..\n\ndx download -f ${regenie_out_dir}/assoc_c*.regenie.gz\nfor pheno in ${pheno_cols//,/ }; do\n    results_out=\"assoc.${pheno}.regenie\"\n    #Title\n    zcat  assoc_c*_${pheno}.regenie.gz | head -1 | sed 's/ /\\t/g' > ${results_out}\n    #Merge\n    for chr in {1..22}; do\n        zcat assoc_c${chr}_${pheno}.regenie.gz | tail -n+2 | sed 's/ /\\t/g' >>  ${results_out}\n    done\n    #Calculate P from LOG10P column, remove values without P and INFO<0.3\n    R -e \"df<-as.data.frame(data.table::fread(\\\"${results_out}\\\"));\\\n          df['P']<-10^(-df['LOG10P']);\\\n          df<-subset(df,LOG10P!='NA');\\\n          df<-subset(df,INFO>0.3);\\\n          data.table::fwrite(df,paste0(\\\"${results_out}\\\",\\\".gz\\\"), sep='\\t', na='NA', quote=F)\"\n    dx upload \"${results_out}.gz\" --path \"${regenie_out_dir}/${results_out}.gz\"\n\ndone   \nrm assoc*regenie*", "metadata": {"trusted": true, "tags": []}, "execution_count": 9, "outputs": [{"name": "stdout", "text": "\u001b[2K[===========================================================>] Completed 117,307,715 of 117,307,715 bytes (100%) /opt/notebooks/assoc_c3_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 129,313,601 of 129,313,601 bytes (100%) /opt/notebooks/assoc_c3_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 127,883,693 of 127,883,693 bytes (100%) /opt/notebooks/assoc_c3_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 153,116,496 of 153,116,496 bytes (100%) /opt/notebooks/assoc_c2_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 140,331,833 of 140,331,833 bytes (100%) /opt/notebooks/assoc_c2_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 154,806,847 of 154,806,847 bytes (100%) /opt/notebooks/assoc_c2_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 104,267,686 of 104,267,686 bytes (100%) /opt/notebooks/assoc_c7_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 95,652,969 of 95,652,969 bytes (100%) /opt/notebooks/assoc_c7_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 105,414,270 of 105,414,270 bytes (100%) /opt/notebooks/assoc_c7_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 141,842,399 of 141,842,399 bytes (100%) /opt/notebooks/assoc_c1_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 143,427,719 of 143,427,719 bytes (100%) /opt/notebooks/assoc_c1_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 130,040,603 of 130,040,603 bytes (100%) /opt/notebooks/assoc_c1_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 80,291,912 of 80,291,912 bytes (100%) /opt/notebooks/assoc_c11_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 88,487,229 of 88,487,229 bytes (100%) /opt/notebooks/assoc_c11_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 87,498,860 of 87,498,860 bytes (100%) /opt/notebooks/assoc_c11_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 125,702,080 of 125,702,080 bytes (100%) /opt/notebooks/assoc_c4_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 124,313,148 of 124,313,148 bytes (100%) /opt/notebooks/assoc_c4_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 114,175,127 of 114,175,127 bytes (100%) /opt/notebooks/assoc_c4_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 111,344,883 of 111,344,883 bytes (100%) /opt/notebooks/assoc_c6_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 101,231,120 of 101,231,120 bytes (100%) /opt/notebooks/assoc_c6_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 110,143,016 of 110,143,016 bytes (100%) /opt/notebooks/assoc_c6_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 114,802,426 of 114,802,426 bytes (100%) /opt/notebooks/assoc_c5_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 116,056,464 of 116,056,464 bytes (100%) /opt/notebooks/assoc_c5_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 105,302,356 of 105,302,356 bytes (100%) /opt/notebooks/assoc_c5_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 98,700,178 of 98,700,178 bytes (100%) /opt/notebooks/assoc_c8_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 99,828,393 of 99,828,393 bytes (100%) /opt/notebooks/assoc_c8_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 90,506,076 of 90,506,076 bytes (100%) /opt/notebooks/assoc_c8_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 78,723,237 of 78,723,237 bytes (100%) /opt/notebooks/assoc_c9_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 77,867,224 of 77,867,224 bytes (100%) /opt/notebooks/assoc_c9_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 71,496,116 of 71,496,116 bytes (100%) /opt/notebooks/assoc_c9_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 89,133,425 of 89,133,425 bytes (100%) /opt/notebooks/assoc_c10_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 80,939,300 of 80,939,300 bytes (100%) /opt/notebooks/assoc_c10_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 88,154,537 of 88,154,537 bytes (100%) /opt/notebooks/assoc_c10_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 77,640,416 of 77,640,416 bytes (100%) /opt/notebooks/assoc_c12_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 84,634,588 of 84,634,588 bytes (100%) /opt/notebooks/assoc_c12_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 85,598,091 of 85,598,091 bytes (100%) /opt/notebooks/assoc_c12_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 59,518,703 of 59,518,703 bytes (100%) /opt/notebooks/assoc_c16_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 58,858,544 of 58,858,544 bytes (100%) /opt/notebooks/assoc_c16_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 53,941,660 of 53,941,660 bytes (100%) /opt/notebooks/assoc_c16_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 62,875,464 of 62,875,464 bytes (100%) /opt/notebooks/assoc_c13_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 63,573,095 of 63,573,095 bytes (100%) /opt/notebooks/assoc_c13_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 57,742,406 of 57,742,406 bytes (100%) /opt/notebooks/assoc_c13_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 57,323,874 of 57,323,874 bytes (100%) /opt/notebooks/assoc_c14_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 57,970,136 of 57,970,136 bytes (100%) /opt/notebooks/assoc_c14_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 52,631,740 of 52,631,740 bytes (100%) /opt/notebooks/assoc_c14_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 48,178,176 of 48,178,176 bytes (100%) /opt/notebooks/assoc_c17_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 52,559,029 of 52,559,029 bytes (100%) /opt/notebooks/assoc_c17_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 53,151,926 of 53,151,926 bytes (100%) /opt/notebooks/assoc_c17_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 52,019,847 of 52,019,847 bytes (100%) /opt/notebooks/assoc_c15_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 52,584,452 of 52,584,452 bytes (100%) /opt/notebooks/assoc_c15_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 47,721,512 of 47,721,512 bytes (100%) /opt/notebooks/assoc_c15_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 50,023,428 of 50,023,428 bytes (100%) /opt/notebooks/assoc_c18_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 45,504,833 of 45,504,833 bytes (100%) /opt/notebooks/assoc_c18_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 49,487,422 of 49,487,422 bytes (100%) /opt/notebooks/assoc_c18_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 41,846,351 of 41,846,351 bytes (100%) /opt/notebooks/assoc_c19_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 38,038,872 of 38,038,872 bytes (100%) /opt/notebooks/assoc_c19_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 41,391,302 of 41,391,302 bytes (100%) /opt/notebooks/assoc_c19_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 41,838,094 of 41,838,094 bytes (100%) /opt/notebooks/assoc_c20_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 41,391,486 of 41,391,486 bytes (100%) /opt/notebooks/assoc_c20_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 37,977,295 of 37,977,295 bytes (100%) /opt/notebooks/assoc_c20_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 23,345,308 of 23,345,308 bytes (100%) /opt/notebooks/assoc_c21_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 23,586,588 of 23,586,588 bytes (100%) /opt/notebooks/assoc_c21_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 21,491,076 of 21,491,076 bytes (100%) /opt/notebooks/assoc_c21_resht_fg.regenie.gzz\n\u001b[2K[===========================================================>] Completed 25,148,786 of 25,148,786 bytes (100%) /opt/notebooks/assoc_c22_resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 25,429,603 of 25,429,603 bytes (100%) /opt/notebooks/assoc_c22_resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 23,081,051 of 23,081,051 bytes (100%) /opt/notebooks/assoc_c22_resht_fg.regenie.gzz\n\ngzip: stdout: Broken pipe\n\nR version 4.2.2 (2022-10-31) -- \"Innocent and Trusting\"\nCopyright (C) 2022 The R Foundation for Statistical Computing\nPlatform: x86_64-pc-linux-gnu (64-bit)\n\nR is free software and comes with ABSOLUTELY NO WARRANTY.\nYou are welcome to redistribute it under certain conditions.\nType 'license()' or 'licence()' for distribution details.\n\nR is a collaborative project with many contributors.\nType 'contributors()' for more information and\n'citation()' on how to cite R or R packages in publications.\n\nType 'demo()' for some demos, 'help()' for on-line help, or\n'help.start()' for an HTML browser interface to help.\nType 'q()' to quit R.\n\n> df<-as.data.frame(data.table::fread(\"assoc.resht_fg.regenie\"));          df['P']<-10^(-df['LOG10P']);          df<-subset(df,LOG10P!='NA');          df<-subset(df,INFO>0.3);          data.table::fwrite(df,paste0(\"assoc.resht_fg.regenie\",\".gz\"), sep='\\t', na='NA', quote=F)\n> \n> \n[===========================================================>] Uploaded 2,062,552,170 of 2,062,552,170 bytes (100%) assoc.resht_fg.regenie.gz\nID                          file-GbBKk8QJFFQpvY1p7bykQXyz\nClass                       file\nProject                     project-GZ62xgQJFFQXbyFk1ZFJkFz0\nFolder                      /data/regenie\nName                        assoc.resht_fg.regenie.gz\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Wed Nov 15 16:06:58 2023\nCreated by                  anirka\n via the job                job-GbBJ008JFFQzxGF3XXPbJ440\nLast modified               Wed Nov 15 16:07:04 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n\ngzip: stdout: Broken pipe\n\nR version 4.2.2 (2022-10-31) -- \"Innocent and Trusting\"\nCopyright (C) 2022 The R Foundation for Statistical Computing\nPlatform: x86_64-pc-linux-gnu (64-bit)\n\nR is free software and comes with ABSOLUTELY NO WARRANTY.\nYou are welcome to redistribute it under certain conditions.\nType 'license()' or 'licence()' for distribution details.\n\nR is a collaborative project with many contributors.\nType 'contributors()' for more information and\n'citation()' on how to cite R or R packages in publications.\n\nType 'demo()' for some demos, 'help()' for on-line help, or\n'help.start()' for an HTML browser interface to help.\nType 'q()' to quit R.\n\n> df<-as.data.frame(data.table::fread(\"assoc.resht_ukb.regenie\"));          df['P']<-10^(-df['LOG10P']);          df<-subset(df,LOG10P!='NA');          df<-subset(df,INFO>0.3);          data.table::fwrite(df,paste0(\"assoc.resht_ukb.regenie\",\".gz\"), sep='\\t', na='NA', quote=F)\n> \n> \n[===========================================================>] Uploaded 2,242,810,892 of 2,242,810,892 bytes (100%) assoc.resht_ukb.regenie.gz\nID                          file-GbBKv1jJFFQqvV3K9225gjPg\nClass                       file\nProject                     project-GZ62xgQJFFQXbyFk1ZFJkFz0\nFolder                      /data/regenie\nName                        assoc.resht_ukb.regenie.gz\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Wed Nov 15 16:12:55 2023\nCreated by                  anirka\n via the job                job-GbBJ008JFFQzxGF3XXPbJ440\nLast modified               Wed Nov 15 16:13:00 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n\ngzip: stdout: Broken pipe\n\nR version 4.2.2 (2022-10-31) -- \"Innocent and Trusting\"\nCopyright (C) 2022 The R Foundation for Statistical Computing\nPlatform: x86_64-pc-linux-gnu (64-bit)\n\nR is free software and comes with ABSOLUTELY NO WARRANTY.\nYou are welcome to redistribute it under certain conditions.\nType 'license()' or 'licence()' for distribution details.\n\nR is a collaborative project with many contributors.\nType 'contributors()' for more information and\n'citation()' on how to cite R or R packages in publications.\n\nType 'demo()' for some demos, 'help()' for on-line help, or\n'help.start()' for an HTML browser interface to help.\nType 'q()' to quit R.\n\n> df<-as.data.frame(data.table::fread(\"assoc.resht_both.regenie\"));          df['P']<-10^(-df['LOG10P']);          df<-subset(df,LOG10P!='NA');          df<-subset(df,INFO>0.3);          data.table::fwrite(df,paste0(\"assoc.resht_both.regenie\",\".gz\"), sep='\\t', na='NA', quote=F)\n> \n> \n[===========================================================>] Uploaded 2,267,139,554 of 2,267,139,554 bytes (100%) assoc.resht_both.regenie.gz\nID                          file-GbBKyv0JFFQf44k8KG7jBYqx\nClass                       file\nProject                     project-GZ62xgQJFFQXbyFk1ZFJkFz0\nFolder                      /data/regenie\nName                        assoc.resht_both.regenie.gz\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Wed Nov 15 16:18:56 2023\nCreated by                  anirka\n via the job                job-GbBJ008JFFQzxGF3XXPbJ440\nLast modified               Wed Nov 15 16:19:01 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "### Draw manhattan and qq plots", "metadata": {}}, {"cell_type": "code", "source": "#This could be run with SAK\n\ndx download -f scripts/draw_manhattan.R\ndx download -f ${regenie_out_dir}/assoc.*.regenie.gz\n#Draw manhattan plots\nfor pheno in ${pheno_cols//,/ }; do\n   Rscript draw_manhattan.R assoc.${pheno}.regenie.gz ${pheno}\ndone\ndx upload *.png --path ${regenie_out_dir}/\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [{"name": "stdout", "text": "\u001b[2K[===========================================================>] Completed 834 of 834 bytes (100%) /opt/notebooks/draw_manhattan.R.R\n\u001b[2K[===========================================================>] Completed 2,267,139,554 of 2,267,139,554 bytes (100%) /opt/notebooks/assoc.resht_both.regenie.gzz\n\u001b[2K[===========================================================>] Completed 2,242,810,892 of 2,242,810,892 bytes (100%) /opt/notebooks/assoc.resht_ukb.regenie.gzz\n\u001b[2K[===========================================================>] Completed 2,062,552,170 of 2,062,552,170 bytes (100%) /opt/notebooks/assoc.resht_fg.regenie.gzz\nLoading required package: data.table\nLoading required package: dplyr\n\nAttaching package: \u2018dplyr\u2019\n\nThe following objects are masked from \u2018package:data.table\u2019:\n\n    between, first, last\n\nThe following objects are masked from \u2018package:stats\u2019:\n\n    filter, lag\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    intersect, setdiff, setequal, union\n\nLoading required package: qqman\n\nFor example usage please run: vignette('qqman')\n\nCitation appreciated but not required:\nTurner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n\nLoading required package: R.utils\nLoading required package: R.oo\nLoading required package: R.methodsS3\nR.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\nR.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.\n\nAttaching package: \u2018R.oo\u2019\n\nThe following object is masked from \u2018package:R.methodsS3\u2019:\n\n    throw\n\nThe following objects are masked from \u2018package:methods\u2019:\n\n    getClasses, getMethods\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    attach, detach, load, save\n\nR.utils v2.12.2 (2022-11-11 22:00:03 UTC) successfully loaded. See ?R.utils for help.\n\nAttaching package: \u2018R.utils\u2019\n\nThe following object is masked from \u2018package:utils\u2019:\n\n    timestamp\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings\n\ndata.table      dplyr      qqman    R.utils \n      TRUE       TRUE       TRUE       TRUE \n\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n          1 \n\u001b[?25h\u001b[?25h\u001b[?25hnull device \n          1 \n\u001b[?25h\u001b[?25hLoading required package: data.table\nLoading required package: dplyr\n\nAttaching package: \u2018dplyr\u2019\n\nThe following objects are masked from \u2018package:data.table\u2019:\n\n    between, first, last\n\nThe following objects are masked from \u2018package:stats\u2019:\n\n    filter, lag\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    intersect, setdiff, setequal, union\n\nLoading required package: qqman\n\nFor example usage please run: vignette('qqman')\n\nCitation appreciated but not required:\nTurner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n\nLoading required package: R.utils\nLoading required package: R.oo\nLoading required package: R.methodsS3\nR.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\nR.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.\n\nAttaching package: \u2018R.oo\u2019\n\nThe following object is masked from \u2018package:R.methodsS3\u2019:\n\n    throw\n\nThe following objects are masked from \u2018package:methods\u2019:\n\n    getClasses, getMethods\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    attach, detach, load, save\n\nR.utils v2.12.2 (2022-11-11 22:00:03 UTC) successfully loaded. See ?R.utils for help.\n\nAttaching package: \u2018R.utils\u2019\n\nThe following object is masked from \u2018package:utils\u2019:\n\n    timestamp\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings\n\ndata.table      dplyr      qqman    R.utils \n      TRUE       TRUE       TRUE       TRUE \n\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n          1 \n\u001b[?25h\u001b[?25h\u001b[?25hnull device \n          1 \n\u001b[?25h\u001b[?25hLoading required package: data.table\nLoading required package: dplyr\n\nAttaching package: \u2018dplyr\u2019\n\nThe following objects are masked from \u2018package:data.table\u2019:\n\n    between, first, last\n\nThe following objects are masked from \u2018package:stats\u2019:\n\n    filter, lag\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    intersect, setdiff, setequal, union\n\nLoading required package: qqman\n\nFor example usage please run: vignette('qqman')\n\nCitation appreciated but not required:\nTurner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n\nLoading required package: R.utils\nLoading required package: R.oo\nLoading required package: R.methodsS3\nR.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\nR.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.\n\nAttaching package: \u2018R.oo\u2019\n\nThe following object is masked from \u2018package:R.methodsS3\u2019:\n\n    throw\n\nThe following objects are masked from \u2018package:methods\u2019:\n\n    getClasses, getMethods\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    attach, detach, load, save\n\nR.utils v2.12.2 (2022-11-11 22:00:03 UTC) successfully loaded. See ?R.utils for help.\n\nAttaching package: \u2018R.utils\u2019\n\nThe following object is masked from \u2018package:utils\u2019:\n\n    timestamp\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings\n\ndata.table      dplyr      qqman    R.utils \n      TRUE       TRUE       TRUE       TRUE \n\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n          1 \n", "output_type": "stream"}]}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}